group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1983,8 +1984,16 @@ private CoordinatorResult<StreamsGroupHeartbeatResult, CoordinatorRecord> stream
         StreamsGroupHeartbeatResponseData response = new StreamsGroupHeartbeatResponseData()
             .setMemberId(updatedMember.memberId())
             .setMemberEpoch(updatedMember.memberEpoch())
-            .setHeartbeatIntervalMs(streamsGroupHeartbeatIntervalMs(groupId))
-            .setPartitionsByUserEndpoint(maybeBuildEndpointToPartitions(group));
+            .setHeartbeatIntervalMs(streamsGroupHeartbeatIntervalMs(groupId));
+
+        if (group.endpointInformationEpoch() > memberEndpointEpochInRequest) {
+            response.setPartitionsByUserEndpoint(maybeBuildEndpointToPartitions(group));
+            response.setEndpointInformationEpoch(group.endpointInformationEpoch());
+        } else {
+            int responseEndpoint = Math.min(group.endpointInformationEpoch(), memberEndpointEpochInRequest);
```

Comment:
We need to account for the `GroupMetadataManager` losing the state of the endpoint epoch, in that case we'll reset the clients to use the GMM endpoint epoch.

Reply:
We do not persist the endpoint-epoch, right? That is the case you are talking about?

But I thought for this case `group.endpointInformationEpoch()` would be `-1` [or whatever value we initialize it] (ie unknown?) -- So `memberEndpointEpochInRequest` should always be larger?

Correct me if I am wrong. The end-to-end control flow, is not totally clear to me atm.

Reply:
I would always respond with `group.endpointInformationEpoch()` and omit this if-block.

If the broker-side epoch is lost, it will be less than then member epoch, in which case we should reset the member epoch to the broker epoch and resend the endpoint information, just in case.